File: Graph/a.articulationPoint.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution {
  public:
    vector<int> articulationPoints(int V, vector<int>adj[]) {
        vector<bool> visited(V , false);
        vector<int> discovery(V);
        vector<int> low(V);
        vector<bool> articulationPoints(V , false);
        
        for(int i = 0 ; i < V ; i++) {
            if(!visited[i]){
                dfs(adj , visited , discovery , low , articulationPoints , i , -1);
            }
        }
        
        int count = 0;
        for(bool aps : articulationPoints){
            if(aps){
                count++;
            }
        }
        if(count == 0){
            vector<int> ans;
            ans.push_back(-1);
            return ans;
        }else {
            vector<int> ans;
            for(int i = 0 ; i < articulationPoints.size() ; i++) {
                if(articulationPoints.at(i) == true){
                    ans.push_back(i);
                }
            }
            return ans;
        }
        
    }
    int time = 0;
    void dfs(vector<int> adj[] , vector<bool>& visited , vector<int>& discovery , vector<int>& low , vector<bool>& articulationPoints , int src , int parent){
        visited[src] = true;
        discovery[src] = low[src] = ++time;
        int child = 0;
        for(int v : adj[src]){
            if(v == parent){ // neighbour parent hai
                continue;
            }else if(visited[v]){ // visited neighour (not parent)
                low[src] = min(discovery[v] , low[src]);
            }else { // unvisited neigbour
                child++;
                dfs(adj , visited , discovery , low , articulationPoints , v , src) ;
                low[src] = min(low[v] , low[src]);
                
                if(parent != -1 && low[v] >= discovery[src]){
                    articulationPoints[src] = true;
                }
            }
        }
        if(parent == -1 && child > 1) {
            articulationPoints[src] = true;
        }
    }
};

// { Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int V, E;
		cin >> V >> E;
		vector<int>adj[V];
		for(int i = 0; i < E; i++){
			int u, v;
			cin >> u >> v;
			adj[u].push_back(v);
			adj[v].push_back(u);
		}
		Solution obj;
		vector<int> ans = obj.articulationPoints(V, adj);
		for(auto i: ans)cout << i << " ";
		cout << "\n";
	}
	return 0;
}  // } Driver Code Ends

File: Graph/a.articulationPoint.java
// { Driver Code Starts
    import java.util.*;
    import java.lang.*;
    import java.io.*;
    class GFG
    {
        public static void main(String[] args) throws IOException
        {
            BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
            int T = Integer.parseInt(br.readLine().trim());
            while(T-->0)
            {
                String[] s = br.readLine().trim().split(" ");
                int V = Integer.parseInt(s[0]);
                int E = Integer.parseInt(s[1]);
                ArrayList<ArrayList<Integer>>adj = new ArrayList<>();
                for(int i = 0; i < V; i++)
                    adj.add(i, new ArrayList<Integer>());
                for(int i = 0; i < E; i++){
                    String[] S = br.readLine().trim().split(" ");
                    int u = Integer.parseInt(S[0]);
                    int v = Integer.parseInt(S[1]);
                    adj.get(u).add(v);
                    adj.get(v).add(u);
                }
                Solution obj = new Solution();
                int[] ans = obj.articulationPoints(V, adj);
                for(int i: ans)
                    System.out.print(i + " ");
                System.out.println();
            }
        }
    }// } Driver Code Ends
    
    
    class Solution
    {
        public int[] articulationPoints(int V, ArrayList<ArrayList<Integer>> adj)
        {
            boolean[] visited = new boolean[V];
            int[] discovery = new int[V];
            int[] low = new int[V];
            boolean[] articulationPoints = new boolean[V];
            for(int i = 0 ; i < V ; i++){
                if(!visited[i]){
                    dfs(adj , visited , discovery , low , articulationPoints , i , -1);
                } 
            }
            int count = 0;
            for(boolean ap : articulationPoints){
                if(ap){
                    count++;
                }
            }
            if(count == 0){
                return new int[] {-1};
            }else{
                int[] res = new int[count];
                int idx = 0;
                for(int i = 0 ; i < articulationPoints.length ; i++){
                    if(articulationPoints[i] == true){
                        res[idx] = i;
                        idx++;
                    }
                }
                 return res;
            }
           
        }
        int time = 0;
        public void dfs(ArrayList<ArrayList<Integer>> adj , boolean[]visited , int[] discovery , int[] low , boolean[] articulationPoints , int u , int p){
            visited[u] = true;
            discovery[u] = low[u] = ++time;
            int child = 0;
            for(int v : adj.get(u)){
                if(v == p){ // neighbour hai par parent hai
                    continue;
                }else if(visited[v]){ // neighbour hai par visited hai(not parent)
                    low[u] =Math.min(discovery[v] , low[u]);
                }else {// unvisited neighbour
                    child++;
                    dfs(adj , visited , discovery , low , articulationPoints , v , u);
                    low[u] = Math.min(low[v] , low[u]);
                    if(p != -1 && low[v] >= discovery[u]){
                        articulationPoints[u] = true;
                    }
                }
                if(p == -1 && child > 1 ){
                    articulationPoints[u] = true;
                }
            }
            
        }
    }